cmd/gaia/init/testnet.go

```diff
@@ -182,67 +200,115 @@ func testnetWithConfig(config *cfg.Config, cdc *codec.Codec) error {
 			return err
 		}
 
-		// Gather gentxs folder
+		// gather gentxs folder
 		err = writeFile(fmt.Sprintf("%v.json", nodeDirName), gentxsDir, txBytes)
 		if err != nil {
 			_ = os.RemoveAll(outDir)
 			return err
 		}
 	}
 
-	// Generate empty genesis.json
+	if err := initGenFiles(cdc, chainID, accs, genFiles); err != nil {
+		return err
+	}
+
+	if err := collectGenFiles(cdc, config, chainID, monikers, nodeIDs, valPubKeys); err != nil {
+		return err
+	}
+
+	fmt.Printf("Successfully initialized %d node directories\n", numValidators)
+	return nil
+}
+
+func initGenFiles(
+	cdc *codec.Codec, chainID string, accs []app.GenesisAccount, genFiles []string,
+) error {
+
+	numValidators := viper.GetInt(flagNumValidators)
```

Comment:
Let's not make our lives more difficult in the future - we should not be calling viper in utility functions - just pass in `numValidators` as an input to `initGenFiles` and only use viper in the highest level function (aka within the command)

Reply:
Yes, I agree. But see my other post [here](https://github.com/cosmos/cosmos-sdk/pull/2692#issuecomment-435933352). This would require even further changes to not only this but all the other auxiliary methods. The function signature would be extremely long.

Being that these are private auxiliary methods contained in a command implementation, I don't see any true harm.

Reply:
sure if it's too much work leave for now I suppose